coalib/files/Fileproxy.py

```diff
@@ -0,0 +1,97 @@
+from coala_utils.decorators import generate_eq
+
+
+@generate_eq("filename")
+class Fileproxy:
+    """
+    The file proxy is a wrapper object to wrap every file. The file proxy is an
+    object that contains properties about the file currently processed (such as
+    the filename). The fileproxy object also protects your file from being
+    modified by a bear. It contains:
+
+        * The filename, this is the starting point for every other attribute.
+        * The content of the file as a string.
+        * The content of the file, line-splitted.
+
+    Note: Other properties can be included later as and when required.
+
+    The equality of the object is checked just based on its filename.
+
+    To initialize a Fileproxy object for a file:
+
+    >>> fp = Fileproxy("./tests/files/FileproxyTestFiles/testfile.txt")
+
+    Now we can get the filename:
+
+    >>> fp.filename
+    './tests/files/FileproxyTestFiles/testfile.txt'
+
+    The number of lines in the file:
+
+    >>> len(fp)
+    1
+
+    Fetch a specific line in the file:
+
+    >>> fp[0]
+    'Old line\\n'
+
+    Iterate through the lines of the file:
+
+    >>> for line in fp:
+    ...     print(line)
+    Old line
+    <BLANKLINE>
+
+    Get contents of the file as a string:
+
+    >>> str(fp)
+    'Old line\\n'
+
+    Get all lines of the file as a list:
+    >>> list(fp)
+    ['Old line\\n']
+    """
+
+    def __init__(self, filename):
+        """
+        Constructs a new fileproxy object.
+
+        :param filename: The name of the file to load.
+        """
+        self.filename = filename
+        with open(self.filename, encoding="utf-8") as filehandle:
+            self.__content = filehandle.read()
+        self.__lines = tuple(self.__content.splitlines(True))
```

Comment:
Having `__lines` and `__content` 'doubles' memory usage compared to the existing implementation.
I strongly suggest that the proxy *attempts* to keep only one of `__lines` or `__content` in memory unless it appears that both are being used.

i.e.

1. `__init__` loads `__content` ; 
2. `__iter__` loads `__lines` and either discards `__content` or switches `__content` to a `weakref` so it can be garbage collected
3. `__str__` loads `__content` from `__lines` if necessary, and doesn't discard `__content`, as both are apparently being used.

Or, a better approach is to only load the file contents when `__iter__` or `__str__` are called, which should reduce startup time , thereby improving responsiveness that the user 'feels'.  However, that approach means the unicode decode handling also needs to be delayed.